go-controller/pkg/node/secondary_node_network_controller.go

```diff
@@ -0,0 +1,82 @@
+package node
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	ovntypes "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog/v2"
+)
+
+// SecondaryNodeNetworkController structure is the object which holds the controls for starting
+// and reacting upon the watched resources (e.g. pods, endpoints) for secondary network
+type SecondaryNodeNetworkController struct {
+	BaseNodeNetworkController
+}
+
+// NewSecondaryNodeNetworkController creates a new OVN controller for creating logical network
+// infrastructure and policy for default l3 network
+func NewSecondaryNodeNetworkController(cnnci *CommonNodeNetworkControllerInfo, netInfo util.NetInfo,
+	netconfInfo util.NetConfInfo) *SecondaryNodeNetworkController {
+	return &SecondaryNodeNetworkController{
+		BaseNodeNetworkController: BaseNodeNetworkController{
+			CommonNodeNetworkControllerInfo: *cnnci,
+			NetConfInfo:                     netconfInfo,
+			NetInfo:                         netInfo,
+			stopChan:                        make(chan struct{}),
+			wg:                              &sync.WaitGroup{},
+		},
+	}
+}
+
+// Start starts the default controller; handles all events and creates all needed logical entities
+func (nc *SecondaryNodeNetworkController) Start(ctx context.Context) error {
+	klog.Infof("Start secondary node network controller of network %s", nc.GetNetworkName())
+
+	isOvnUpEnabled := atomic.LoadInt32(&nc.atomicOvnUpEnabled) > 0
+
+	if err := nc.nodeUpgrade(ctx, isOvnUpEnabled, nil, nil, ""); err != nil {
```

Comment:
so each controller runs its own upgrade, in the future it would be nice if it was a single upgrade goroutine that other controllers are subscribed to. But we can do that later.